src/jit/morph.cpp

```diff
@@ -11416,6 +11416,20 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                 }
             }
 
+            // If gtOp1 is a GT_FIELD, we need to pass down the mac if
+            // its parent is GT_ADDR, since the address of the field
+            // is part of an ongoing address computation. Otherwise
+            // op1 represents the value of the field and so any address
+            // calculations it does are in a new context.
+            if ((op1->gtOper == GT_FIELD) && (tree->gtOper != GT_ADDR))
```

Comment:
I'd think that (somewhat guessing here):

  1. If the pointer-valued child FIELD is loading a managed pointer, then we're not required to insert a null check (otherwise we'd be adding an additional null check at virtually every Span indexer, no?), and instead can rely on the null check have been performed before storing the value into the field.
  2. If the pointer-valued child FIELD is loading an object pointer, then us seeing an ADD here instead of an ADDR (FIELD obj, f) here indicates that the input was some sort of unsafe code that doesn't require a null check.

In any case, passing the address context on down to the recursive call that processes the 2nd FIELD could only lead to null checking the address that the FIELD loads from, which of course wouldn't serve the purpose of null checking the loaded value to which the indefinite offset is added.